tests/integ/test_dataframe.py

```diff
@@ -2437,10 +2437,66 @@ def test_describe(session):
     assert "invalid identifier" in str(ex_info)
 
 
+def test_overwrite_vs_truncate(session):
+    table_name = Utils.random_name_for_temp_object(TempObjectType.TABLE)
+    df = session.create_dataframe([(1, 2), (3, 4)]).toDF("a", "b")
+    # we create the table
+    df.write.save_as_table(table_name, table_type="temp")
+    original_schema = df.schema
+    # we will use overwrite. Because it recreates the table the schema will change
+    df = session.create_dataframe([(5, 5), (6, 6)]).toDF("b", "a")
+    df.write.save_as_table(table_name, mode="overwrite", table_type="temp")
+    new_schema = df.schema
+    # same length
```

Comment:
This test is correct but longer and slightly harder to follow. Let's simplify it to capture the main difference between overwrite and truncate

```
tmp_table_name1 = Utils.random_name_for_temp_object(TempObjectType.TABLE)
df1 = session.create_dataframe([(1, 2), (3, 4)], schema=["a", "b"])
df2 = session.create_dataframe([(1, 2, 3), (4, 5, 6)], schema=["a", "b", "c"])

df1.write.save_as_table(tmp_table_name, table_type="temp")

# truncate preserves old schema
with pytest.raises(SnowparkSQLException, match="encountered schema mismatch"):
    df2.write.save_as_table(tmp_table_name, mode="truncate", table_type="temp")


# overwrite drops old schema
df2.write.save_as_table(tmp_table_name, mode="overwrite", table_type="temp")
Utils.check_answer(session.table(tmp_table_name), [Row(1, 2, 3), Row(4, 5, 6)])